src/ivectorbin/merge-vads.cc

```diff
@@ -0,0 +1,131 @@
+// ivectorbin/merge-vads.cc
+
+// Copyright  2015 David Snyder
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "matrix/kaldi-matrix.h"
+#include "util/stl-utils.h"
+
+using namespace kaldi;
+using kaldi::int32;
+
+void PrepareMap(const std::string map_rxfilename,
+  unordered_map<std::pair<int32, int32>, int32, PairHasher<int32> > *map) {
+  Input map_input(map_rxfilename);
+
+  // If a map file isn't specified, provide an obvious mapping.  The
+  // following mapping assumes "0" corresponds to nonspeech and "1"
+  // corresponds to speech. The combination of two VAD decisions only
+  // results in a decision of speech if both input frames are
+  // classified as speech.
+  if (map_rxfilename.empty()) {
+    (*map)[std::pair<int32, int32>(0, 0)] = 0;
+    (*map)[std::pair<int32, int32>(0, 1)] = 0;
+    (*map)[std::pair<int32, int32>(1, 0)] = 0;
+    (*map)[std::pair<int32, int32>(1, 1)] = 1;
+  } else {
+    std::string line;
+    while (std::getline(map_input.Stream(), line)) {
+      std::vector<std::string> fields;
+      SplitStringToVector(line, " \t\n\r", true, &fields);
+      if (fields.size() != 3) {
+        KALDI_ERR << "Bad line. Expected three fields, got: "
+                  << line;
+      }
+      int32 label1 = std::atoi(fields[0].c_str()),
+            label2 = std::atoi(fields[1].c_str()),
+            result_label = std::atoi(fields[2].c_str());
+      (*map)[std::pair<int32, int32>(label1, label2)] = result_label;
+    }
+  }
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    const char *usage =
+      "This program merges two Vector archives of frame-level voice\n"
+      "activity decisions. By default, the program assumes that the\n"
+      "input vectors consist of floats that are 0.0 if a frame is judged\n"
+      "as nonspeech and 1.0 if it is considered speech. The default\n"
+      "produces an output Vector that is 1.0 if both inputs are 1.0\n"
```

Comment:
I think it's better to leave them as floats, because then they are
compatible with the output of ali-to-weights (which can be configured to
output 0.0 and 1.0).
Dan


On Fri, Dec 4, 2015 at 12:34 PM, david-ryan-snyder <notifications@github.com
> wrote:

> In src/ivectorbin/merge-vads.cc
> <https://github.com/kaldi-asr/kaldi/pull/354#discussion_r46708290>:
>
> > +      int32 label1 = std::atoi(fields[0].c_str()),
> > +            label2 = std::atoi(fields[1].c_str()),
> > +            result_label = std::atoi(fields[2].c_str());
> > +      (*map)[std::pair<int32, int32>(label1, label2)] = result_label;
> > +    }
> > +  }
> > +}
> > +
> > +int main(int argc, char *argv[]) {
> > +  try {
> > +    const char *usage =
> > +      "This program merges two Vector archives of frame-level voice\n"
> > +      "activity decisions. By default, the program assumes that the\n"
> > +      "input vectors consist of floats that are 0.0 if a frame is judged\n"
> > +      "as nonspeech and 1.0 if it is considered speech. The default\n"
> > +      "produces an output Vector that is 1.0 if both inputs are 1.0\n"
>
> I think the interpretation of these vectors is *not* the same as weights.
> This is more like the output of compute-vad (in that binary, floats are
> either 0.0 or 1.0, there is no logical interpretation of 0.5). Even though
> it is stored as a float, it is basically handled as an integer.
>
> In order to make these binaries compatible with compute-vad, I'm also
> storing the VAD decisions of compute-vad-from-likes and merge-vads as
> archives of float Vectors. But now that I think about it, I don't really
> see any advantage to doing it this way. There's currently no case where we
> use a frame-level VAD where it makes sense to have nonintegral VAD
> decisions. Would you mind if we change compute-vad and select-voiced-frames
> to work with Int32Vectors instead? Of course, I would also modify the new
> VAD programs in the same way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/354/files#r46708290>.
>